### ISSUE TITLE
using `import` with non-local paths yields unexpected results with C++ code generation.

### ISSUE THREAD

**Issue by stefanseefeld** [protocolbuffers/protobuf#4737]
I'm trying to include a proto file from another directory using `import "foo/bar.proto";`, which results in code generation that somehow incorporates the "foo" directory name in the generated code (but only in references, not actual declared names).
This is entirely unexpected (as well as undocumented as far as I can see), as there also is the "module" spec which is used to wrap declarations into namespaces.

In a nutshell, using `import "foo/bar.proto";` seems entirely unusable, so I'd like to understand a) whether that's intended and b) how to work around it.

**Comment by acozzette** [comment]
@stefanseefeld I think the issue here is that you have to be consistent about how you invoke protoc when you build foo/bar.proto and the other proto file which imports it. If one file imports "foo/bar.proto" then make sure you pass "foo/bar.proto" to `protoc` when you build it (e.g. don't cd to foo/ and then call protoc on bar.proto).

**Comment by stefanseefeld** [comment]
@acozzette Yes, I understand that what you suggest would prevent the error, but it seems quite counter-intuitive to require this. Isn't the entire purpose of the `-I` option to provide a similar flexibility as C/C++ preprocessing ?
Or, to approach it from the other end: what is the purpose of `protoc` handling it the way it does ? Would anything break if it started ignoring the `dirname` of the imported file name ?

In any case, this is at least undocumented, but I'd go further and declare it a (user interface) bug.

**Comment by xfxyjwf** [comment]
How the protoc currently works is by design. It's a very very common misunderstanding to think that "-I" flag for protoc works the same way as it is for C++ compiler. It isn't and is not designed to work that way. Protoc uses the "-I" option to determine the import path of every .proto file (regardless whether the .proto file is imported or not, it will have a globally unique import path) and generates code that is dependent on that path. For example, that import path is embedded in the generated descriptor data which is later used by DescriptorPool and APIs like DescriptorPool::FindFileByName() relies on that import path as the key to look up descriptors in the entire binary. Ignoring the dirname or not, protoc needs a unique name for every .proto file, and if dirname is ignored, protoc will have to require every .proto file to have a globally unique file name and that's not going to happen.

I agree this is counter-intuitive though. I see too often that users complain about generated code not working because they are not aware of the significance of the import path. Some will even suggest we change the generated code to make it compile (e.g., https://github.com/google/protobuf/issues/1491) but it won't work because of the "unique global import path for every .proto file" requirement. A generated code change will not allow you to do "import ./foo.proto" in one place and "import ../foo.proto" in some other place because the generated code will have embedded descriptor data referencing the same file with different names and that will fail at runtime.

We definitely need better documentation for this. I actually started some documentation for this before but never got the time to finish/publish it...

**Comment by stefanseefeld** [comment]
On 2018-06-18 08:51 PM, Feng Xiao wrote:
>
> How the protoc currently works is by design. It's a very very common 
> misunderstanding to think that "-I" flag for protoc works the same way 
> as it is for C++ compiler. It isn't and is not designed to work that way.
>
OK, then please document that.

> Protoc uses the "-I" option to determine the import path of every 
> .proto file (regardless whether the .proto file is imported or not, it 
> will have a globally unique import path) and generates code that is 
> dependent on that path. For example, that import path is embedded in 
> the generated descriptor data which is later used by DescriptorPool 
> and APIs like DescriptorPool::FindFileByName() relies on that import 
> path as the key to look up descriptors in the entire binary. Ignoring 
> the dirname or not, protoc needs a unique name for every .proto file, 
> and if dirname is ignored, protoc will have to require every .proto 
> file to have a globally unique file name and that's not going to happen.
>

<speculating>
It could work just the way C/C++ preprocessing works, i.e. it could use 
the import search path strictly for looking up proto files, 
irrespectively of their content. In fact, my surprise stems in part from 
the fact that there also is the "module" spec that generates namespaces 
(in the C++ mapping), and thus could be used to disambiguate.
</speculating>

> I agree this is counter-intuitive though. I see too often that users 
> complain about generated code not working because they are not aware 
> of the significance of the import path. Some will even suggest we 
> change the generated code to make it compile (e.g., #1491 
> <https://github.com/google/protobuf/issues/1491>) but it won't work 
> because of the "unique global import path for every .proto file" 
> requirement. A generated code change will not allow you to do "import 
> ./foo.proto" in one place and "import ../foo.proto" in some other 
> place because the generated code will have embedded descriptor data 
> referencing the same file with different names and that will fail at 
> runtime.
>

What I still don't understand is why that requirement exists. Couldn't 
`protoc` simply reuse the established semantics (and idioms) associated 
with C/C++ preprocessing, so a processed file is uniquely identified, no 
matter by what name I include it (which of course also requires some 
form of header guarding (#pragma once could be implicit)).

> We definitely need better documentation for this. I actually started 
> some documentation for this before but never got the time to 
> finish/publish it...
>

OK. In any case, despite my complaining, I'm very grateful for the 
protobuf framework, so thanks for making this available ! :-)


Stefan

--

       ...ich hab' noch einen Koffer in Berlin...
